### ISSUE TITLE
anyone could explain to me about stream documentation ?

### ISSUE THREAD

**Issue by laptopmutia** [stedolan/jq#1984]
https://stedolan.github.io/jq/manual/#Streaming

what is this doc means?

Streaming
With the --stream option jq can parse input texts in a streaming fashion, allowing jq programs to start processing large JSON texts immediately rather than after the parse completes. If you have a single JSON text that is 1GB in size, streaming it will allow you to process it much more quickly.

However, streaming isn’t easy to deal with as the jq program will have [<path>, <leaf-value>] (and a few other forms) as inputs.

Several builtins are provided to make handling streams easier.

The examples below use the streamed form of [0,[1]], which is [[0],0],[[1,0],1],[[1,0]],[[1]].

Streaming forms include [<path>, <leaf-value>] (to indicate any scalar value, empty array, or empty object), and [<path>] (to indicate the end of an array or object). Future versions of jq run with --stream and -seq may output additional forms such as ["error message"] when an input text fails to parse.

what is this [path, leaf-value] and what is with 0,[1] ?

from the faq the example is this

```
jq -n '[{foo:"bar"},{foo:"baz"}]' | jq -cn --stream 'fromstream(1|truncate_stream(inputs))'
```

I don't really understand this what is this `1|truncate_stream(inputs)` means? I'm really confused

**Comment by jszabo98** [comment]
I agree, this is hard to understand.  I was just looking to process a big json file in powershell and got interested.  Here's the streaming version of that json.  I'm doing this in powershell.  Add an echo in front for bash.  --stream turns the json into a list of the paths to the values.  And then there's kind of an end marker for every array or object.  This post might help:  http://subtxt.in/library-data/2016/03/28/json_stream_jq

```json
'[{"foo":"bar"},{"foo":"baz"}]' | jq --stream -c
[[0,"foo"],"bar"]
[[0,"foo"]]
[[1,"foo"],"baz"]
[[1,"foo"]]
[[1]]
```

Here it is after the "parent folder' in the paths is deleted by truncate_stream.  You could put 0 instead of 1 to see it before the change.

```json
'[{"foo":"bar"},{"foo":"baz"}]' | jq -cn --stream '1|truncate_stream(inputs)'            
[["foo"],"bar"]
[["foo"]]
[["foo"],"baz"]
[["foo"]]
```
And then fromstream() turns it back into json:

```json
'[{"foo":"bar"},{"foo":"baz"}]' | jq -cn --stream 'fromstream(1|truncate_stream(inputs))'
{"foo":"bar"}
{"foo":"baz"}
```

**Comment by austinarbor** [comment]
> This post might help: http://subtxt.in/library-data/2016/03/28/json_stream_jq

In the post , they run 
`zcat < all.json.gz | jq --stream "fromstream(1|truncate_stream(inputs))" | jq "select(any(._source.sourceResource; .type=="sound"))"` 

If i'm not mistaken, isn't it just taking the source, turning it into a stream, then right back into a json object? Do you know how this is more performant (memory-wise) than just running the `jq "select(any(._source.sourceResource; .type=="sound"))"` directly?

**Comment by pkoppstein** [comment]
@austinarbor - The --stream option parses the JSON incrementally and is thus quite different from the regular parser. See https://github.com/stedolan/jq/wiki/FAQ#streaming-json-parser

**Comment by austinarbor** [comment]
@pkoppstein is `jq --stream "fromstream(1|truncate_stream(inputs))"` constructing the json objects one at a time and piping them over to the 2nd jq invocation to process as each object is constructed?

**Comment by pkoppstein** [comment]
@austinarbor - There is only one invocation of jq in your example, so I am not sure what you mean by "2nd jq invocation". If you want to see what's going on, consider using `debug`. By the way, if you're going to use `inputs`, you will almost certainly want to use the -n command-line option. E.g. you could compare `truncate_stream(debug)` with  `truncate_stream(inputs|debug)`

**Comment by austinarbor** [comment]
@pkoppstein in the blog post @jszabo98 linked, the command is 
`zcat < all.json.gz | jq --stream "fromstream(1|truncate_stream(inputs))" | jq "select(any(._source.sourceResource; .type=="sound"))"`

the `jq --stream "fromstream(1|truncate_stream(inputs))"` is piped into `jq "select(any(._source.sourceResource; .type=="sound"))"`, so the filtering is done after the streamed json is converted back into a json object. I'm just wondering how this is more efficient, since I would assume that the the 2nd jq invocation loses any benefit the first streaming invocation had

**Comment by pkoppstein** [comment]
@austinarbor - Yes, it would be silly to invoke jq twice in that way.  That's not how to using the --stream option. If you want to use it properly, you'll have to gain some understanding e.g. by reading up on it, and perhaps using `debug`.

**Comment by austinarbor** [comment]
@pkoppstein thanks for clarifying - I understand the docs, I just wanted to **confirm** my suspicion that the blog's command was not correct

**Comment by pkoppstein** [comment]
@ustinarbor - Please note that I revised my earlier response.
